### PR TITLE
🎨Computational backend: DV-2 computational scheduler becomes replicable (🗃️🚨)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

This PR heavily refactors the director-v2 internal computational scheduler by using tools to allow multiple director-v2 replicas to be able to share the load of scheduling computational pipelines.

As reminder the computational scheduler in the director-v2 is responsible for:
1. analyzing a project's computational pipeline and to find which nodes need to be (re-)run when the user starts a computation,
2. scheduling computational jobs according to the computational pipeline (what can run and what shall wait, retrieving results/errors and transmitting progress) by communicating with the shared Dask cluster or with private Dask clusters

Until this PR, replicating the director-v2 would also duplicate network calls and end in wasted resources.

This PR aims to make the replication of the director-v2 more efficient by:
- having a Manager that periodically schedules (5s) the pipelines, relying on the Database information directly, exclusively (e.g. only 1 director-v2 replica can do it at a time - using distributed locks),
- these tasks are published in a RabbitMQ shared queue named `simcore.services.director-v2.scheduling`,
- each director-v2 replica contains _COMPUTATIONAL_BACKEND_SCHEDULING_CONCURRENCY_ workers that _apply_ the scheduling and communicate with the deeper backend (e.g. Dask), this allows to share the load between the different workers, so in effect every new replica of the director-v2 shall relieve the other ones,
- `COMPUTATIONAL_BACKEND_SCHEDULING_CONCURRENCY`  is currently hard-coded to _50_. Some testing will be necessary to see whether that is too low or too high. That is why this is not at the moment an ENV variable. It will be converted if necessary.
- 🗃️: `comp_runs` table is upgraded to contain new nullable _scheduled_ and _processed_ columns, this is used to keep track of when a pipeline was _scheduled_ by the manager, and when the worker has _processed_ it.
- 🗃️: `comp_runs` table is upgraded to contain only timezone-enabled timestamps
- 🗃️: `comp_tasks` table is upgraded to contain only timezone-enabled timestamps

### Schematic
```mermaid
---
config:
  theme: mc
  layout: dagre
  look: handDrawn
---
flowchart LR
 subgraph s1["Director-v2.1"]
        n1["Manager:<br>5s: Schedule all pipelines"]
        n3["Worker1"]
        n4["Worker2"]
        n5["WorkerN"]
        n6["schedule pipeline1"]
        n7["schedule pipeline2"]
        n8["schedule pipeline3"]
  end
 subgraph s2["Cluster-UserX"]
        n9["Dask-Scheduler"]
        n11["Dask-Worker(s)"]
  end
 subgraph s3["Cluster-UserY"]
        n14["Dask-Scheduler"]
        n15["Dask-Worker(s)"]
  end
 subgraph s4["Cluster-UserZ"]
        n19["Dask-Scheduler"]
        n20["Dask-Worker(s)"]
  end
 subgraph s5["Director-v2.2"]
        n21["Manager:<br>5s: Schedule all pipelines"]
        n22["Worker1"]
        n23["Worker2"]
        n24["WorkerN"]
        n25["schedule pipeline4"]
        n26["schedule pipeline5"]
        n27["schedule pipeline6"]
  end
 subgraph s6["Cluster-UserA"]
        n90["Dask-Scheduler"]
        n91["Dask-Worker(s)"]
  end
 subgraph s7["Cluster-UserB"]
        n92["Dask-Scheduler"]
        n93["Dask-Worker(s)"]
  end
 subgraph s8["Cluster-UserC"]
        n94["Dask-Scheduler"]
        n95["Dask-Worker(s)"]
  end
    n1 ==> n2["RabbitMQ"]
    n2 --> n3 & n4 & n5 & n22 & n23 & n24
    n3 --> n6
    n4 --> n7
    n5 --> n8
    n6 --> n9
    n7 --> n14
    n8 --> n19
    n22 --> n25
    n23 --> n26
    n24 --> n27
    n21 ==> n2
    n25 --> n90
    n26 --> n92
    n27 --> n94
    n21@{ shape: rect}
    n22@{ shape: rect}
    n23@{ shape: rect}
    n24@{ shape: rect}
    n25@{ shape: rect}
    n26@{ shape: rect}
    n27@{ shape: rect}
    n2@{ shape: cyl}
    style n1 stroke:#D50000
    style n6 stroke:#D50000
    style n7 stroke:#D50000
    style n8 stroke:#D50000
    style n21 stroke:#D50000,stroke-width:1px,stroke-dasharray: 1
    style n25 stroke:#D50000
    style n26 stroke:#D50000
    style n27 stroke:#D50000


```
Legend:
- in **$${\color{red}red}$$** are blocks that run _exclusively_ via Redis locks

## Dev-ops checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/_worker.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    rabbitmq_client = get_rabbitmq_client(app)
+    app.state.scheduler_worker_consumers = await asyncio.gather(
+        *(
+            rabbitmq_client.subscribe(
+                SchedulePipelineRabbitMessage.get_channel_name(),
+                functools.partial(_handle_apply_distributed_schedule, app),
+                exclusive_queue=False,
+            )
+            for _ in range(
+                app_settings.DIRECTOR_V2_COMPUTATIONAL_BACKEND.COMPUTATIONAL_BACKEND_SCHEDULING_CONCURRENCY
```
Comment (matusdrobuliak66) [inline services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/_worker.py:77]: Observation: Currently we have: `qos_value = 1 if exclusive_queue is False else _DEFAULT_PREFETCH_VALUE` where _DEFAULT_PREFETCH_VALUE = 10.

Alternatively, you could have parameterized the qos_value and used a single consumer that consumes more messages, and then the consumer might have spawned multiple workers.

File: services/director-v2/src/simcore_service_director_v2/models/comp_runs.py [hunk 2]
```diff
@@ -51,6 +55,8 @@ class CompRunsAtDB(BaseModel):
     cancelled: datetime.datetime | None
     metadata: RunMetadataDict = RunMetadataDict()
     use_on_demand_clusters: bool
+    scheduled: datetime.datetime | None
+    processed: datetime.datetime | None
```
Comment (matusdrobuliak66) [inline services/director-v2/src/simcore_service_director_v2/models/comp_runs.py:59]: MINOR: `scheduled_at`, `processed_at`

**Review by matusdrobuliak66 — APPROVED** [review]
Very nice, thanks!

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/_worker.py (hunk 1 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/_worker.py:77]: so setting a qos of 10 means that each worker would in effect take 10 messages, but the handler would still trigger sequentially. In this way one would need to re-publish the failed task in the rabbitmq from the worker again.
I decided to go that way cause it was a bit simpler. But we'll see if that becomes an issue.

File: services/director-v2/src/simcore_service_director_v2/core/settings.py [hunk 3]
```diff
@@ -72,6 +74,10 @@ class ComputationalBackendSettings(BaseCustomSettings):
     COMPUTATIONAL_BACKEND_ENABLED: bool = Field(
         default=True,
     )
+    COMPUTATIONAL_BACKEND_SCHEDULING_CONCURRENCY: PositiveInt = Field(
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/core/settings.py:77]: maybe ask expose this in the `.env-devel` and relative OPS branches so we can tune it easily on production

**Review by GitHK — APPROVED** [review]

File: services/director-v2/src/simcore_service_director_v2/core/settings.py (hunk 3 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/core/settings.py:77]: precisely I do not want to expose it at the moment. I want to play in master if needed.

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_runs.py [hunk 4]
```diff
@@ -55,24 +60,74 @@ async def get(
             return CompRunsAtDB.model_validate(row)
 
     async def list(
-        self, filter_by_state: set[RunningState] | None = None
+        self,
+        *,
+        filter_by_state: set[RunningState] | None = None,
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_runs.py:65]: An interesting use-case is when you want to associate `None` to some other state. For instance
https://github.com/itisfoundation/osparc-simcore/blob/5732a12e07e63d5ce55010ede9b9ab543bb9b278/services/web/server/src/simcore_service_webserver/folders/_folders_db.py#L322-L326

File: packages/postgres-database/src/simcore_postgres_database/models/comp_tasks.py [hunk 5]
```diff
@@ -77,9 +78,15 @@ class NodeClass(enum.Enum):
         doc="current progress of the task if available",
     ),
     # utc timestamps for submission/start/end
-    sa.Column("submit", sa.DateTime, doc="UTC timestamp for task submission"),
-    sa.Column("start", sa.DateTime, doc="UTC timestamp when task started"),
-    sa.Column("end", sa.DateTime, doc="UTC timestamp for task completion"),
+    sa.Column(
+        "submit", sa.DateTime(timezone=True), doc="UTC timestamp for task submission"
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/postgres-database/src/simcore_postgres_database/models/comp_tasks.py:82]: These entries are older than you 🤣 
I prefer not to add even more noise in this PR.

File: packages/service-library/src/servicelib/exception_utils.py [hunk 6]
```diff
@@ -65,3 +68,33 @@ def else_reset(self) -> None:
         """error no longer occurs reset tracking"""
         self._first_exception_skip = None
         self._failure_counter = 0
+
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def silence_exceptions(
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/service-library/src/servicelib/exception_utils.py]: yes. but what I want is something I can use with tasks that is generic.
ok these context decorators are interesting indeed, but I am not sure I see the added value in this case. let's discuss that tomorrow.

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_runs.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_runs.py:65]: well to be honest and as I already said, I don't really see the point of using something else than None as a sentinel but for Rest APIs actually.. and only for PATCH calls actually.
not sure I see the point of enforcing this everywhere else.

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_runs.py [hunk 7]
```diff
@@ -117,7 +172,16 @@ async def create(
                 row = await result.first()
                 return CompRunsAtDB.model_validate(row)
         except ForeignKeyViolation as exc:
-            raise ClusterNotFoundError(cluster_id=cluster_id) from exc
+            message = exc.args[0]
+            match message:
+                case s if "users" in s and "user_id" in s:
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_runs.py:177]: yes, chatty generated that and it is usually weak and use old school stuff. I can change that.

File: packages/service-library/src/servicelib/exception_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/service-library/src/servicelib/exception_utils.py]: ok tested, but there is sadly no async suppress

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/_scheduler_base.py [hunk 8]
```diff
@@ -676,32 +531,20 @@ async def _process_completed_tasks(
...
         user_id: UserID,
         project_id: ProjectID,
-        iteration: PositiveInt,
-        wake_up_callback: Callable[[], None],
+        iteration: Iteration,
     ) -> None:
+        """schedules a pipeline for a given user, project and iteration.
+
+        Arguments:
+            wake_up_callback -- a callback function that is called in a separate thread everytime a pipeline node is completed
```
Comment (bisgaard-itis) [inline services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/_scheduler_base.py:546]: I don't think this `wake_up_callback` doc belongs here. Or am I misunderstanding something?

**Review by bisgaard-itis — COMMENTED** [review]
Very interesting! Looking forward to seeing multiple director-v2 replicas. Unfortunately I am not deep enough in the details to give a more thorough review, but I did discover one little thing. 🙂

**Pull request merged**